[LHeng1] iP

LHeng1 opened the pull request:
# Avalon
> “People die if they are killed.” – Emiya Shirou [(totally legit source)](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

![The Last Sleep Of Arthur](https://github.com/nus-cs2103-AY2324S1/ip/assets/110957690/5b203a6d-aa35-4ddd-a47c-5a8f72740533)

King Arthur is getting bored being rested in Avalon after the Battle of Camlann, and now attempts to reconnect with the modern world by helping you remember things you need to do. It's,

- **Text-Based** ~~Arthur has yet to learn GUI~~
- **Interactive** ~~Arthur has longed for companionship~~
- **SUPER FAST to use!!!** ~~Arthur needs friends...~~

All you need to do is,
1. Download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
1. Double-click it.
1. Input your tasks.
1. Let Mr.Excalibur ⚔️⚔️ manage your tasks for you!!

<details>
    <summary> Best Part: </summary> 

    IT'S FREE!!!!

</details>

Features:
- [X] Task Creation
- [X] Task List
- [X] Task Marking
- [X] Task Deletion
- [X] Task Search Filter
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
- [ ] Free Excalibur (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public static void main(String[] args) {
        new Avalon("src/main/data/Avalon.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

LHeng1 committed:
Level-0

LHeng1 committed:
Level-1

LHeng1 committed:
Level-2-Add-List

LHeng1 committed:
Level-3-MarkAsDone

LHeng1 committed:
Level-4-ToDo-Event-Deadline

LHeng1 committed:
TextUiTesting

LHeng1 committed:
Level-5-HandleErrors

LHeng1 committed:
Level-6-Delete-WO-Collections

LHeng1 committed:
A-Enums-UseEnums-Changeable

LHeng1 committed:
Change of Directory

LHeng1 committed:
Change of directory

vansh284 commented on src/main/java/Avalon.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
```
Perhaps write a header comment for the class to improve code readability? I noticed that this is missing in other classes/public methods that were added in this PR as well.

vansh284 commented on src/main/java/Avalon.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.Scanner;
+public class Avalon {
+    public static void main(String[] args) {
+        Task[] tasks = new Task[100];
```
I like how you named the array using the plural form.

vansh284 commented on src/main/java/Avalon.java:
```diff
@@ -0,0 +1,136 @@
+        Task[] tasks = new Task[100];
+        int taskCount = 0;
+        Scanner scanner = new Scanner(System.in);
+        //greet
```
Do we have separate logical units here (as suggested by the comment)? In that case, a blank line might improve readability here. There were other instances in the code base as well which could benefit from such improved readability.

vansh284 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+    enum TaskType {
+        TODO, DEADLINE, EVENT
+    }
+    protected String description;
```
There has been consistent use of camelCase for the variables in the PR which I like!

vansh284 reviewed (commented):
Overall, the PR seems to mostly comply with the given coding standard, just some nits to fix. Another consideration: grouping the classes into packages may help other developers more easily understand the code.

LHeng1 committed:
Add save function & optimize collections

LHeng1 committed:
Merge branch 'branch-Level-7'

LHeng1 committed:
Teach chatbot date and time

LHeng1 committed:
Merge branch 'branch-Level-8'

LHeng1 committed:
Setup ui class

LHeng1 committed:
Setup storage class and improve tasklist class

LHeng1 committed:
Refactor code to be more OOP

LHeng1 committed:
Merge branch 'branch-A-MoreOOP'

LHeng1 committed:
Organize classes into packages

LHeng1 committed:
Merge branch 'branch-A-Packages'

LHeng1 committed:
Merge branch 'add-gradle-support'

LHeng1 committed:
Setup gradle for automation

LHeng1 committed:
Setup gradle to run unit tests using JUnit

teozhengyang commented on src/main/java/avalon/Avalon.java:
```diff
@@ -0,0 +1,114 @@
+                        }
+                        description = parts[0];
+                        String by = parts[1];
+
+                        Deadline deadline = new Deadline(description, by);
+                        tasks.addTask(deadline);
+                        ui.showAddTaskMessage(tasks);
+                        break;
+                    case "event":
+                        parts = userInput.substring(6).split(" /from | /to ");
+                        if (parts.length != 3) {
+                            throw new IllegalArgumentException("Please provide a description, a starting time, and an ending time (use /from and /to).");
+                        }
+                        description = parts[0];
+                        String from = parts[1];
+                        String to = parts[2];
+
+                        Event event = new Event(description, from, to);
+                        tasks.addTask(event);
+                        ui.showAddTaskMessage(tasks);
+                        break;
+                    case "delete":
+                        taskIndex = Integer.parseInt(userInput.substring(7)) - 1;
+                        if (taskIndex >= 0 && taskIndex < tasks.size()) {
+                            Task deletedTask = tasks.get(taskIndex);
+                            tasks.removeTask(taskIndex);
+                            ui.showDeleteTaskMessage(tasks, deletedTask);
+                        } else {
+                            throw new IllegalArgumentException("Invalid task number to be deleted.");
+                        }
+                        break;
+                    default:
+                        throw new IllegalArgumentException("I humbly apologize, but thy words remain a mystery to me...");
+                }
+            } catch (IllegalArgumentException e) {
+                System.out.println("☹ OOPS!!! " + e.getMessage());
+            }
+        }
+
+    }
```
Better to avoid long methods and keep it within 30 LoC.

teozhengyang reviewed (commented):
Overall LGTM. Apart from the long method that can be spilt into a few methods, everything is alright.

LHeng1 committed:
Add basic JUnit tests to 2 classes

LHeng1 committed:
Merge branch 'branch-A-JUnit'

LHeng1 committed:
Configure IntelliJ IDE to follow code style

LHeng1 committed:
Fix layout of code

LHeng1 committed:
Add JavaDoc comments to the codebase

LHeng1 committed:
Add find keyword function

LHeng1 committed:
Merge branch 'branch-Level-9'

LHeng1 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/avalon/Avalon.java

LHeng1 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/avalon/Avalon.java
#	src/main/java/avalon/Task.java
#	src/main/java/avalon/TaskList.java
#	src/main/java/avalon/TaskParser.java
#	src/main/java/avalon/Ui.java

LHeng1 committed:
Create JavaFX hello world program

LHeng1 committed:
Create simple GUI for Avalon

LHeng1 committed:
Complete GUI tutorial

LHeng1 committed:
Merge branch 'branch-Level-10'

LHeng1 committed:
Fix checkstyle errors in main and test files

LHeng1 committed:
Merge branch 'A-CheckStyle'

LHeng1 committed:
Used varargs for search keyword function

LHeng1 committed:
Merge branch 'branch-A-Varargs'

LHeng1 committed:
Enhance code quality and structure

This commit improves the overall code quality and structure of
the Avalon project. This commit address several aspects:

Code Duplication: There were instances of code duplication, especially
in the various command classes. This commit eliminates duplicate
code by extracting common behaviors into separate methods and
classes.

Modularity: Refactor the codebase to enhance modularity
and maintainability. Extract related code into separate
classes and packages, following better organization practices.

Error Handling: Introduce a custom exception class,
AvalonException, to handle errors more gracefully.
This will make it easier to diagnose and resolve issues in the future.

UI Improvement: Enhance user interface (UI) for better
user interaction.

Standardization: Follows a standardized package
structure, making it more organized and comprehensible.

Dead code: Delete dead code as it is redundant and is not used in any
classes.

LHeng1 committed:
Improve code quality and add documentation

Few classes in the codebase lacks documentation,
which makes it hard to understand and maintain.
Additionally, there is a hardcoded file path in the Avalon class,
which is not a good practice and makes the code less flexible.

Let's address these issues by doing the following:

Add JavaDocs to classes and methods throughout the codebase,
providing clear documentation for each component's purpose and usage.

Replace the hardcoded file path in the Avalon class with a constant
field, making it more readable and
allowing for easier future changes in file paths.

LHeng1 committed:
Add assertions to codebase

Added assertions to document critical assumptions in the codebase.

Introduce assertions at key points in the Storage class to verify
the ability to write to the tasks file before saving tasks.

Introduce assertions at key points in both Parser classes to verify
input and output are in the correct format.

LHeng1 committed:
Merge pull request #2 from LHeng1/branch-A-CodeQuality

Improve code quality

LHeng1 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/avalon/utility/DateTimeParser.java

LHeng1 committed:
Merge pull request #3 from LHeng1/branch-A-Assertions

Add assertions to codebase

LHeng1 committed:
Implement set priority functionality

LHeng1 committed:
Merge branch 'branch-C-Priority'

LHeng1 committed:
Improve GUI and personality of chatbot.

LHeng1 committed:
Merge branch 'branch-A-BetterGui'

LHeng1 committed:
Add JUnit test cases to one more class

LHeng1 committed:
Merge branch 'branch-A-MoreTesting'

LHeng1 committed:
Add a brief user guide

LHeng1 committed:
Update README.md

LHeng1 committed:
Revert "Update README.md"

This reverts commit 8eafd81a1aacbe0729d4e46d96ef1afa4645b788.

LHeng1 committed:
Update README file

LHeng1 committed:
Rename UI file

LHeng1 committed:
Create JAR file

LHeng1 committed:
Fix file error

LHeng1 committed:
Rename file

LHeng1 committed:
Correct file name
